### ISSUE TITLE
Incorrect Result?

### ISSUE THREAD

**Issue by jinyang175** [HarvardPL/formulog#12]
Hi guys, 

Consider the following program: 

```
input a(i32).
input d(i32).
@bottomup
output b(i32).
@topdown
output c(i32, i32).
@topdown
output ppp(i32).
@topdown
output q(i32).
a(0). a(1). a(2).
d(0). d(2). d(4).
ppp(X) :- a(X).
b(X) :- d(X).
c(X, Y) :- b(X), ppp(X), b(Y).
q(X) :- c(X, X).
output ok.
ok :- q(2).

output e__( i32, i32)
output f__(i32, i32)
output g__(i32, i32)   (* ------------ Output relation ------------ *)
output h__(i32)
output i__(string, i32)

e__(Var_b, Var_a) :- c(Var_a, Var_b).
f__(Var_d, Var_c) :- e__(_, Var_d), a(Var_c), !b(Var_c), ppp(_), e__(_, _).
f__(Var_b, Var_b) :- e__(Var_a, Var_b), !e__(Var_a, Var_b).
g__(Var_a, Var_c) :- d(Var_a), f__(_, Var_c).
h__(Var_g) :- f__(Var_g, Var_f), f__(Var_g, _), f__(Var_g, Var_g), q(_), h__(Var_f).
```

I get the following result for the relation `g__`
```
g__(0, 1)
g__(2, 1)
g__(4, 1)
```

If however I add a new rule `f__(Var_a, Var_b) :- f__(Var_a, Var_b).`, the result for `g__` should not change. But if I run the following new program:
```
input a(i32).
input d(i32).
@bottomup
output b(i32).
@topdown
output c(i32, i32).
@topdown
output ppp(i32).
@topdown
output q(i32).
a(0). a(1). a(2).
d(0). d(2). d(4).
ppp(X) :- a(X).
b(X) :- d(X).
c(X, Y) :- b(X), ppp(X), b(Y).
q(X) :- c(X, X).
output ok.
ok :- q(2).

output e__( i32, i32)
output f__(i32, i32)
output g__(i32, i32)   (* ------------ Output relation ------------ *)
output h__(i32)
output i__(string, i32)

e__(Var_b, Var_a) :- c(Var_a, Var_b).
f__(Var_d, Var_c) :- e__(_, Var_d), a(Var_c), !b(Var_c), ppp(_), e__(_, _).
f__(Var_b, Var_b) :- e__(Var_a, Var_b), !e__(Var_a, Var_b).
f__(Var_a, Var_b) :- f__(Var_a, Var_b).  (* ------------ New Rule ------------ *)
g__(Var_a, Var_c) :- d(Var_a), f__(_, Var_c).
h__(Var_g) :- f__(Var_g, Var_f), f__(Var_g, _), f__(Var_g, Var_g), q(_), h__(Var_f).
```

I get: 
```
g__(0, 1)
g__(0, 2)
g__(2, 1)
g__(2, 2)
g__(4, 1)
g__(4, 2)
```

Formulog version I am using: 508c7688bbe826a6951cc8c3e70026e57548e044 
Operating System: Debian GNU/Linux 10 (buster)
Kernel: Linux 5.4.163.1.amd64-smp

**Comment by aaronbembenek** [comment]
Thanks for raising the issue! That definitely sounds like it could be a bug. However, I'm having trouble reproducing it - right now I am getting just these values in both cases:
```
g__(0, 1)
g__(2, 1)
g__(4, 1)
```
Is this issue happening intermittently or consistently for you? And do you happen to know what the correct output is for the relation `g__`?

**Comment by jinyang175** [comment]
Hi @aaronbembenek,
For relation `g__` the correct output should indeed be
```
g__(0, 1)
g__(2, 1)
g__(4, 1)
```
Yes I am consistently getting different results for the two programs. I am getting the following outputs.
For the first program: 
```
Parsing...
Finished parsing (0.412s)
Type checking...
Finished type checking (0.015s)
Rewriting and validating...
Finished rewriting and validating (0.361s)
Evaluating...
Finished evaluating (0.015s)
Extensional database:
a(0)
a(1)
a(2)
d(0)
d(2)
d(4)

Intensional database:
b(0)
b(2)
b(4)
c(0, 0)
c(0, 2)
c(0, 4)
c(2, 0)
c(2, 2)
c(2, 4)
e__(0, 0)
e__(0, 2)
e__(2, 0)
e__(2, 2)
e__(4, 0)
e__(4, 2)
f__(0, 1)
f__(2, 1)
g__(0, 1)
g__(2, 1)
g__(4, 1)
input_c_bb(2, 2)
input_c_ff
input_ppp_b(0)
input_ppp_b(2)
input_ppp_b(4)
input_ppp_f
input_q_b(2)
input_q_f
ok
p_b(0)
p_b(2)
p_b(4)
p_c(0, 0)
p_c(0, 2)
p_c(0, 4)
p_c(2, 0)
p_c(2, 2)
p_c(2, 4)
p_e__(0, 0)
p_e__(0, 2)
p_e__(2, 0)
p_e__(2, 2)
p_e__(4, 0)
p_e__(4, 2)
p_f__(0, 0)
p_f__(0, 1)
p_f__(0, 2)
p_f__(2, 0)
p_f__(2, 1)
p_f__(2, 2)
p_g__(0, 0)
p_g__(0, 1)
p_g__(0, 2)
p_g__(2, 0)
p_g__(2, 1)
p_g__(2, 2)
p_g__(4, 0)
p_g__(4, 1)
p_g__(4, 2)
p_ok
p_ppp(0)
p_ppp(1)
p_ppp(2)
p_q(0)
p_q(2)
ppp(0)
ppp(1)
ppp(2)
q(0)
q(2)
sup_0_0
sup_10_0
sup_11_0(2, 2)
sup_2_0
sup_4_0(0, 0)
sup_4_0(0, 2)
sup_4_0(1, 0)
sup_4_0(1, 2)
sup_4_0(2, 0)
sup_4_0(2, 2)
sup_6_0(0, 0)
sup_6_0(0, 2)
sup_6_0(1, 0)
sup_6_0(1, 2)
sup_6_0(2, 0)
sup_6_0(2, 2)
sup_7_0(2)
sup_8_0(0)
sup_8_0(2)
sup_8_0(4)
```

For the second program: 
```
Parsing...
Finished parsing (0.198s)
Type checking...
Finished type checking (0.015s)
Rewriting and validating...
Finished rewriting and validating (0.582s)
Evaluating...
Finished evaluating (0.012s)
Extensional database:
a(0)
a(1)
a(2)
d(0)
d(2)
d(4)

Intensional database:
b(0)
b(2)
b(4)
c(0, 0)
c(0, 2)
c(0, 4)
c(2, 0)
c(2, 2)
c(2, 4)
e__(0, 0)
e__(0, 2)
e__(2, 0)
e__(2, 2)
e__(4, 0)
e__(4, 2)
f__(0, 1)
f__(2, 1)
f__(2, 2)
g__(0, 1)
g__(0, 2)
g__(2, 1)
g__(2, 2)
g__(4, 1)
g__(4, 2)
input_c_bb(2, 2)
input_c_ff
input_ppp_b(0)
input_ppp_b(2)
input_ppp_b(4)
input_ppp_f
input_q_b(2)
input_q_f
ok
ppp(0)
ppp(1)
ppp(2)
q(0)
q(2)
sup_0_0
sup_11_0
sup_12_0(2, 2)
sup_2_0
sup_5_0(1, 0)
sup_5_0(1, 2)
sup_7_0(1, 2)
sup_7_0(2, 2)
sup_8_0(2)
sup_9_0(0)
sup_9_0(2)
sup_9_0(4)
```

Are we getting different results? 
I can also try to find some more examples if you are not able to reproduce this. 
Are you running these examples on a mac?

**Comment by aaronbembenek** [comment]
I am indeed running things on a Mac, so let me try running on a Linux machine closer to what you have and get back to you.

**Comment by aaronbembenek** [comment]
I tried running things on a Debian Buster container, and I still wasn't able to reproduce the behavior you're seeing. However, from looking at the output, I have an inkling of what might be happening. Could you share the output when you run the second program with the property `-DdebugMst`, as in `java -DdebugMst -jar formulog.jar ...`?

**Comment by jinyang175** [comment]
Oh that is strange. 
Here is the output when I run the second program with `-DdebugMst`:
```
Parsing...
Finished parsing (0.487s)
Type checking...
Finished type checking (0.016s)
Rewriting and validating...
Adorning rules...
--0--
ok :- q_b(2).
--1--
b(X) :- d(X).
--2--
e__(Var_b, Var_a) :- c_ff(Var_a, Var_b).
--3--
f__(Var_a, Var_b) :- f__(Var_a, Var_b).
--4--
f__(Var_b, Var_b) :-
	e__(Var_a, Var_b),
	!e__(Var_a, Var_b).
--5--
f__(Var_d, Var_c) :-
	e__($0, Var_d),
	a(Var_c),
	!b(Var_c),
	ppp_f($1),
	e__($2, $3).
--6--
g__(Var_a, Var_c) :-
	d(Var_a),
	f__($4, Var_c).
--7--
h__(Var_g) :-
	f__(Var_g, Var_f),
	f__(Var_g, $5),
	f__(Var_g, Var_g),
	q_f($6),
	h__(Var_f).
--8--
q_b(X) :- c_bb(X, X).
--9--
c_ff(X, Y) :-
	b(X),
	ppp_b(X),
	b(Y).
--10--
ppp_f(X) :- a(X).
--11--
q_f(X) :- c_ff(X, X).
--12--
c_bb(X, Y) :-
	b(X),
	ppp_b(X),
	b(Y).
--13--
ppp_b(X) :- a(X).
Rewritten rules...
b(X) :- d(X).
c(X, Y) :-
	sup_12_0(X, Y),
	ppp(X),
	b(Y).
c(X, Y) :-
	sup_9_0(X),
	ppp(X),
	b(Y).
e__(Var_b, Var_a) :-
	sup_2_0,
	c(Var_a, Var_b).
f__(Var_a, Var_b) :- f__(Var_a, Var_b).
f__(Var_d, Var_c) :-
	sup_5_0(Var_c, Var_d),
	ppp($1),
	e__($2, $3).
f__(Var_b, Var_b) :-
	e__(Var_a, Var_b),
	!e__(Var_a, Var_b).
g__(Var_a, Var_c) :-
	d(Var_a),
	f__($4, Var_c).
h__(Var_g) :-
	sup_7_0(Var_f, Var_g),
	q($6),
	h__(Var_f).
input_c_bb(X, X) :- sup_8_0(X).
input_c_ff :- sup_2_0.
input_c_ff :- sup_11_0.
input_ppp_b(X) :- sup_9_0(X).
input_ppp_b(X) :- sup_12_0(X, Y).
input_ppp_f :- sup_5_0(Var_c, Var_d).
input_q_b(2) :- sup_0_0.
input_q_f :- sup_7_0(Var_f, Var_g).
ok :-
	sup_0_0,
	q(2).
ppp(X) :-
	input_ppp_b(X),
	a(X).
ppp(X) :-
	input_ppp_f,
	a(X).
q(X) :-
	sup_8_0(X),
	c(X, X).
q(X) :-
	sup_11_0,
	c(X, X).
sup_0_0 :- true = true.
sup_11_0 :- input_q_f.
sup_12_0(X, Y) :-
	input_c_bb(X, Y),
	b(X).
sup_2_0 :- true = true.
sup_5_0(Var_c, Var_d) :-
	e__($0, Var_d),
	a(Var_c),
	!b(Var_c).
sup_7_0(Var_f, Var_g) :-
	f__(Var_g, Var_f),
	f__(Var_g, $5),
	f__(Var_g, Var_g).
sup_8_0(X) :- input_q_b(X).
sup_9_0(X) :-
	input_c_ff,
	b(X).
Finished rewriting and validating (0.944s)
Evaluating...
Finished evaluating (0.011s)
Extensional database:
a(0)
a(1)
a(2)
d(0)
d(2)
d(4)

Intensional database:
b(0)
b(2)
b(4)
c(0, 0)
c(0, 2)
c(0, 4)
c(2, 0)
c(2, 2)
c(2, 4)
e__(0, 0)
e__(0, 2)
e__(2, 0)
e__(2, 2)
e__(4, 0)
e__(4, 2)
f__(0, 1)
f__(2, 1)
f__(2, 2)
g__(0, 1)
g__(0, 2)
g__(2, 1)
g__(2, 2)
g__(4, 1)
g__(4, 2)
input_c_bb(2, 2)
input_c_ff
input_ppp_b(0)
input_ppp_b(2)
input_ppp_b(4)
input_ppp_f
input_q_b(2)
input_q_f
ok
ppp(0)
ppp(1)
ppp(2)
q(0)
q(2)
sup_0_0
sup_11_0
sup_12_0(2, 2)
sup_2_0
sup_5_0(1, 0)
sup_5_0(1, 2)
sup_7_0(1, 2)
sup_7_0(2, 2)
sup_8_0(2)
sup_9_0(0)
sup_9_0(2)
sup_9_0(4)
```

**Comment by aaronbembenek** [comment]
I think I've found the bug, which was in the routine to determine whether a program contains non-stratified negation, and which I've fixed in 9a6e50a6f95dff6520c2148ee78b51b0fb8cc772. Given that I cannot reproduce the behavior of your test cases myself, do you mind checking if this fixes things on your end? Thanks!

**Comment by jinyang175** [comment]
That indeed fixed the issue :) 
Thanks for the fix. Closing the issue.
